### PR TITLE
[#10325] Google blocking E2E login on production server 

### LINKED ISSUE #10325: Google blocking E2E login on production server
- **Environment**: `master` branch at commit dd9977ec150a49c93040004c4eb034a43fe0180a

**Steps to reproduce**
Run any E2E test on production server

**Expected behaviour**
E2E tests should be able to run

**Actual behaviour**
E2E tests unable to log in as Google blocks log in via automation

**Additional info**
https://stackoverflow.com/questions/60117232/selenium-google-login-block

Error on sign in:
sign-in-problem

### PR DESCRIPTION
**Pull request by jtankw3**
Fixes #10325

Bypass google block by logging in manually and using that user data to run automated test.
Thus, we can skip over the login page as user is already logged in.

However, this only allows one test to be run at a time on Chrome due to 'user data directory is already in use' error.

**Outline of Solution**

- Add firefox profile or chrome user data folders to webdriver to include user data in test.

### CODE REVIEW

File: src/e2e/java/teammates/e2e/cases/e2e/BaseE2ETestCase.java [hunk 1]
```diff
@@ -82,6 +82,12 @@ protected static Url createLocalUrl(String testFileName) throws IOException {
      * Logs in a page using admin credentials (i.e. in masquerade mode).
      */
     protected <T extends AppPage> T loginAdminToPage(AppUrl url, Class<T> typeOfPage) {
+        // When not using dev server, Google blocks log in by automation.
+        // To log in, log in manually to teammates in your browser before running e2e tests.
+        // Refer to teammates.e2e.pageobjects.Browser for more information.
+        if (!TestProperties.isDevServer()) {
+            return null;
```
Comment (xpdavid) [inline src/e2e/java/teammates/e2e/cases/e2e/BaseE2ETestCase.java:89]: Will this result in NPE?

File: src/e2e/java/teammates/e2e/cases/e2e/BaseE2ETestCase.java [hunk 2]
```diff
@@ -82,6 +82,12 @@ protected static Url createLocalUrl(String testFileName) throws IOException {
      * Logs in a page using admin credentials (i.e. in masquerade mode).
      */
     protected <T extends AppPage> T loginAdminToPage(AppUrl url, Class<T> typeOfPage) {
```
Comment (xpdavid) [inline src/e2e/java/teammates/e2e/cases/e2e/BaseE2ETestCase.java:84]: We should also delete code for google login?

File: src/e2e/java/teammates/e2e/pageobjects/Browser.java [hunk 3]
```diff
@@ -126,10 +127,19 @@ private WebDriver createWebDriver() {
...
             System.setProperty("webdriver.gecko.driver", TestProperties.GECKODRIVER_PATH);
 
+            FirefoxProfile profile;
+            if (TestProperties.isDevServer()) {
+                profile = new FirefoxProfile();
+            } else {
+                // Get user data from browser to bypass google blocking automated log in.
+                // Log in manually to teammates to use that log in data for e2e tests.
+                ProfilesIni profileIni = new ProfilesIni();
+                profile = profileIni.getProfile(TestProperties.FIREFOX_PROFILE_NAME);
```
Comment (xpdavid) [inline src/e2e/java/teammates/e2e/pageobjects/Browser.java:137]: I think we should do check the existing of the profile here. It's more user friendly.

File: src/e2e/java/teammates/e2e/pageobjects/Browser.java [hunk 4]
```diff
@@ -146,6 +156,12 @@ private WebDriver createWebDriver() {
 
             ChromeOptions options = new ChromeOptions();
             options.addArguments("--allow-file-access-from-files");
+            if (!TestProperties.isDevServer()) {
+                // Get user data from browser to bypass google blocking automated log in.
+                // Log in manually to teammates to use that log in data for e2e tests.
+                options.addArguments("user-data-dir=" + TestProperties.CHROME_USER_DATA_PATH);
```
Comment (xpdavid) [inline src/e2e/java/teammates/e2e/pageobjects/Browser.java:162]: Same here.

File: src/e2e/java/teammates/e2e/cases/e2e/BaseE2ETestCase.java [hunk 5]
```diff
@@ -82,6 +82,12 @@ protected static Url createLocalUrl(String testFileName) throws IOException {
...
+        // When not using dev server, Google blocks log in by automation.
+        // To log in, log in manually to teammates in your browser before running e2e tests.
+        // Refer to teammates.e2e.pageobjects.Browser for more information.
```
Comment (xpdavid) [inline src/e2e/java/teammates/e2e/cases/e2e/BaseE2ETestCase.java:87]: I think it is more appropriate to throw exception with this message (and result in test fail). It is more explicit than a random NPE error where we need to trace down it to this line.

File: src/e2e/java/teammates/e2e/cases/e2e/BaseE2ETestCase.java (hunk 1 above)
Comment (jtankw3) [reply src/e2e/java/teammates/e2e/cases/e2e/BaseE2ETestCase.java]: I have changed it to return the AppPage of the url it should navigate to instead.

File: src/e2e/java/teammates/e2e/cases/e2e/BaseE2ETestCase.java (hunk 5 above)
Comment (jtankw3) [reply src/e2e/java/teammates/e2e/cases/e2e/BaseE2ETestCase.java:88]: I was actually intending to skip the login process when the production server is used and not to cause failure.

File: src/e2e/java/teammates/e2e/pageobjects/GoogleLoginPage.java [hunk 6]
```diff
@@ -1,38 +1,15 @@
...
-import org.openqa.selenium.support.ui.ExpectedConditions;
-
-import com.google.common.base.Preconditions;
-
-import teammates.e2e.util.TestProperties;
-
 /**
  * Page Object Model for the official Google Accounts login page.
  */
 public class GoogleLoginPage extends LoginPage {
```
Comment (xpdavid) [inline src/e2e/java/teammates/e2e/pageobjects/GoogleLoginPage.java:6]: Actually this entire page can be deleted

Comment (wkurniawan07) [reply src/e2e/java/teammates/e2e/pageobjects/GoogleLoginPage.java:6]: Let's keep it for now. I have a feeling we might be using this again some time; if I'm wrong we can always delete it later.

**Pull request merged**